RecoPixelVertexing/PixelTrackFitting/plugins/PixelTrackReconstructionGPU.cu

```diff
@@ -1,124 +1,210 @@
-#include <Eigen/Core>
-#include <Eigen/Eigenvalues>
-
 #include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
 #include "PixelTrackReconstructionGPU.h"
 
+#include <Eigen/Core>
+#include <Eigen/Eigenvalues>
+
 #ifndef GPU_DEBUG
 #define GPU_DEBUG 0
 #endif // GPU_DEBUG
 
 using namespace Eigen;
 
-__global__
-void KernelFullFitAllHits(float * hits_and_covariances,
-    int hits_in_fit,
-    int cumulative_size,
-    double B,
-    Rfit::helix_fit * results) {
-  // Reshape Eigen components from hits_and_covariances, using proper thread and block indices
-  // Perform the fit
-  // Store the results in the proper vector, using again correct indices
+__global__ void
+KernelFastFitAllHits(float *hits_and_covariances, int hits_in_fit,
+                     int cumulative_size, float B, Rfit::helix_fit *results,
+                     Rfit::Matrix3xNd *hits, Rfit::Matrix3Nd *hits_cov,
+                     Rfit::circle_fit *circle_fit, Vector4d *fast_fit,
+                     Rfit::line_fit *line_fit) {
+
+  // Reshape Eigen components from hits_and_covariances, using proper thread and
+  // block indices Perform the fit Store the results in the proper vector, using
+  // again correct indices
 
   // Loop for hits_in_fit times:
   //   first 3 are the points
   //   the rest is the covariance matrix, 3x3
   int start = (blockIdx.x * blockDim.x + threadIdx.x) * hits_in_fit * 12;
-  int helix_start = (blockIdx.x * blockDim.x + threadIdx.x); 
+  int helix_start = (blockIdx.x * blockDim.x + threadIdx.x);
   if (start >= cumulative_size) {
     return;
   }
 
-#if GPU_DEBUG
-  printf("BlockDim.x: %d, BlockIdx.x: %d, threadIdx.x: %d, start: %d, cumulative_size: %d\n",
-      blockDim.x, blockIdx.x, threadIdx.x, start, cumulative_size);
-#endif
+  if (GPU_DEBUG) {
```

Comment:
Can you use `#ifdef` instead of `if (...)` ?

Reply:
C++17 `if constexpr` might be nicer, but we do not have C++17 in CUDA device code :(